### ISSUE TITLE
[Question] [Enhancement] Download and process external images

### ISSUE THREAD

**Issue by kaboomdigital** [getzola/zola#1052]
Would it be possible to somehow process an external image with the `resize_image` function?

The resize image function uses a local path for the image to process. But I would like to process an image from an Airtable base. (So Airtable can be used as a sort of CMS). I'm trying to figure out a way how to get the image from the URL and how to add it to Zola so it can be processed.

Just pointing the `path` argument to an external URL and Zola downloading and processing it would be a nice feature to add maybe. But for now; perhaps someone can point me in the direction to look for a sexy solution?

Running a separate process which downloads everything before starting Zola would be an option. But perhaps there is a way which is more integrated into the build process?

Thanks for helping out. :-)

**Comment by kaboomdigital** [comment]
The documentation for Tera has a section for Rust [functions](https://tera.netlify.app/docs/#functions). Would it be possible to write a function that fetches some images? Or is this template-functionality meant for other logic?

**Comment by RealOrangeOne** [comment]
Using a function is definitely the right way to go about implementing this. 

I suspect what you'd want to do is check whether the `path` argument looks like a URL, and if so download the file to some temp location (temp deterministic, so it could be reused on rebuilds), then pass the downloaded path back into the `resize_image` method as a path. 

The source for `resize_image` is https://github.com/getzola/zola/blob/master/components/templates/src/global_fns/mod.rs#L149

The source for the `load_data` zola function might also be of useful reference: https://github.com/getzola/zola/blob/master/components/templates/src/global_fns/load_data.rs, as that also downloads remote data

**Comment by kaboomdigital** [comment]
@RealOrangeOne Thanks for the input! :-) 

I started implementing everything in a separate script because I also want to dynamically generate pages based on content from the same data-source as the images. I think it would be definitely useful to have an implementation of the `resize_images` that can process external URLs, so I'll leave this issue open for now. (example use case: images stored in a CMS)

**Comment by mjp4** [comment]
Hi, I'm keen to fix this. I have a prototype working where I've added "jpeg" as a format to load_data() which then allows me to download an image and then resize it.

```
  {% set remote_image = load_data(url="http://my_images.com/img1.jpg", format="jpeg") %}
  {% set image = resize_image(path=remote_image.static_path, height=100, width=100, format="webp", quality=10) %}
  {% if image %}
    <img src="{{ image.url }}">
  {% endif %}
```

I have opted to extend `load_data` to handle binary files (and returning an object with the destination file path) as it has a lot of useful functionality, but other alternatives could include pulling the guts out of `load_data.call()` and sharing it between the existing function and a new `get_remote.call()` function.

I'm loathe to further overload `resize_image` with the additional arguments needed, but that is another option.

I'm currently mirroring the behaviour of `resize_image` and downloading to a subdirectory of `static/` (currently called `remote_data/`) where the filename is the hashed value of the input arguments.

I can see a use case for either wanting the image file to appear in the final site, or only as an interim file (but definitely want to cache them), so I think putting them in static is best.

I don't know whether there's much other call for more of a pipeline of image processing, but if so then it might be worth extending `imageproc` to fully support that and add this as a step.

@Keats Do you have any opinions? Shall I make a PR? Do you have any particular requirements for contributions?

**Comment by Keats** [comment]
I'm wondering whether this should be part of zola or some other pipeline. It would have to download the file in the static folder, but how do you diff it across runs? Do you just download the file everytime it's running? How do you know the file is the same otherwise?

**Comment by teotwaki** [comment]
@Keats I would tend to agree that this is a content issue, not zola related. Might I suggest closing the issue?

**Comment by ebkalderon** [comment]
Ran into a case where this functionality would be very nice to have. I have a shortcode which embeds a Mastodon post into a Zola page. I would like to download any image/video media attachments associated with the post rather than hotlinking to the external resource. Self-hosting the media as a static site resource would provide two benefits: (1) persistence in case the original post or social media site is ever deleted, and (2) the asset can be preprocessed in a responsive manner to improve site-loading performance.
